File: tests/algorithms/x/Clojure/backtracking/combination_sum.clj
(ns main (:refer-clojure :exclude [backtrack combination_sum]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare backtrack combination_sum)

(def ^:dynamic backtrack_i nil)

(def ^:dynamic backtrack_new_path nil)

(def ^:dynamic backtrack_result nil)

(def ^:dynamic backtrack_value nil)

(def ^:dynamic combination_sum_path nil)

(def ^:dynamic combination_sum_result nil)

(defn backtrack [backtrack_candidates backtrack_start backtrack_target backtrack_path backtrack_result_p]
  (binding [backtrack_i nil backtrack_new_path nil backtrack_result nil backtrack_value nil] (try (do (set! backtrack_result backtrack_result_p) (when (= backtrack_target 0) (throw (ex-info "return" {:v (conj backtrack_result backtrack_path)}))) (set! backtrack_i backtrack_start) (while (< backtrack_i (count backtrack_candidates)) (do (set! backtrack_value (nth backtrack_candidates backtrack_i)) (when (<= backtrack_value backtrack_target) (do (set! backtrack_new_path (conj backtrack_path backtrack_value)) (set! backtrack_result (backtrack backtrack_candidates backtrack_i (- backtrack_target backtrack_value) backtrack_new_path backtrack_result)))) (set! backtrack_i (+ backtrack_i 1)))) (throw (ex-info "return" {:v backtrack_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn combination_sum [combination_sum_candidates combination_sum_target]
  (binding [combination_sum_path nil combination_sum_result nil] (try (do (set! combination_sum_path []) (set! combination_sum_result []) (throw (ex-info "return" {:v (backtrack combination_sum_candidates 0 combination_sum_target combination_sum_path combination_sum_result)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (str (combination_sum [2 3 5] 8)))
      (println (str (combination_sum [2 3 6 7] 7)))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Kotlin/backtracking/combination_sum.kt
var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun backtrack(candidates: MutableList<Int>, start: Int, target: Int, path: MutableList<Int>, result: MutableList<MutableList<Int>>): MutableList<MutableList<Int>> {
    var result: MutableList<MutableList<Int>> = result
    if (target == 0) {
        return run { val _tmp = result.toMutableList(); _tmp.add(path); _tmp }
    }
    var i: Int = start
    while (i < candidates.size) {
        var value: Int = candidates[i]!!
        if (value <= target) {
            var new_path = run { val _tmp = path.toMutableList(); _tmp.add(value); _tmp }
            result = backtrack(candidates, i, target - value, (new_path as MutableList<Int>), result)
        }
        i = i + 1
    }
    return result
}

fun combination_sum(candidates: MutableList<Int>, target: Int): MutableList<MutableList<Int>> {
    var path: MutableList<Int> = mutableListOf<Int>()
    var result: MutableList<MutableList<Int>> = mutableListOf<MutableList<Int>>()
    return backtrack(candidates, 0, target, path, result)
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(combination_sum(mutableListOf(2, 3, 5), 8).toString())
        println(combination_sum(mutableListOf(2, 3, 6, 7), 7).toString())
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
